### PR TITLE
[feature](nereids): polish property deriver enforcer job

### PR DESCRIPTION
**Pull request by jackwener**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Polish property deriver enforcer job

- More scenarios were considered
- Add RequestPropertyDeriver to support more `parent property drive`

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (No Need)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    public static List<EqualTo> getEqualTo(PhysicalHashJoin<Plan, Plan> join) {
+        List<SlotReference> leftSlots = join.left().getOutput().stream().map(slot -> (SlotReference) slot)
+                .collect(Collectors.toList());
+        List<SlotReference> rightSlots = join.right().getOutput().stream().map(slot -> (SlotReference) slot)
+                .collect(Collectors.toList());
+
+        List<EqualTo> eqConjuncts = Lists.newArrayList();
+
+
+        if (!join.getCondition().isPresent()) {
```
Comment (924060929) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:59]: you can move this condition to the start of this function

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            requestPropertyToChildren.add(propertiesForShuffle);
+        }
+        if (!JoinUtils.onlyShuffle(hashJoin)) {
+            requestPropertyToChildren.add(propertiesForBroadcast);
+        }
+
+        return null;
+    }
+
+    protected static List<PhysicalProperties> computeShuffleJoinRequiredProperties(
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:95]: this function do not used anymore

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+import org.apache.doris.nereids.util.Utils;
+
+import com.google.common.collect.Lists;
+
+import java.util.List;
+
+/**
+ * Used for parent property drive.
+ */
+public class RequestPropertyDeriver extends PlanVisitor<Void, PlanContext> {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:39]: do we need to add other node later?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecAny.java [hunk 4]
```diff
@@ -18,8 +18,16 @@
 package org.apache.doris.nereids.properties;
 
 /**
- * Describe random distribution.
+ * Data can be anywhere on the segments (required only).
  */
-public class RandomDistributionDesc extends DistributionSpec {
+public class DistributionSpecAny extends DistributionSpec {
 
+    public DistributionSpecAny() {
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecAny.java:25]: This is learn GPORCA name.
I think its advantage is `DistributionXXXX` is put together in files list.
image

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java (hunk 3 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:39]: Yes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecAny.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecAny.java:25]: Use singleton is better

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpec.java [hunk 5]
```diff
@@ -21,52 +21,38 @@
...
-    public void setDataPartition(DataPartition dataPartition) {
-        this.dataPartition = dataPartition;
-    }
+    public abstract boolean meet(DistributionSpec other);
 
     public GroupExpression addEnforcer(Group child) {
         PhysicalDistribution distribution = new PhysicalDistribution(
-                new DistributionSpec(dataPartition), child.getLogicalProperties(), new GroupPlan(child));
+                this,
+                child.getLogicalProperties(),
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpec.java:43]: good catch, but I haven't figured out how to do it yet

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpec.java [hunk 6]
```diff
@@ -21,52 +21,38 @@
...
-    }
-
-    public DataPartition getDataPartition() {
-        return dataPartition;
-    }
-
-    public void setDataPartition(DataPartition dataPartition) {
-        this.dataPartition = dataPartition;
-    }
+    public abstract boolean meet(DistributionSpec other);
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpec.java:38]: I add comment, and have renamed it

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecGather.java [hunk 7]
```diff
@@ -20,9 +20,14 @@
 /**
  * Re-shuffle.
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecGather.java:21]: Yes, it's history comment, and it's outdated

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecAny.java [hunk 8]
```diff
@@ -18,8 +18,16 @@
...
+        super();
+    }
+
+    @Override
+    public boolean meet(DistributionSpec other) {
+        return other instanceof DistributionSpecAny;
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecAny.java:31]: super's function's comment is wrong, I have fixed it.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpec.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpec.java:43]: Add a todo

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java (hunk 2 above)
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/RequestPropertyDeriver.java:95]: Maybe we we will it in the future

File: fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+        // DistributionSpecHash spec = (DistributionSpecHash) other;
+        return false;
+    }
+
+    /**
+     * information of hash distribution.
+     * tableId relate with {@link org.apache.doris.catalog.Table}
+     */
+    public static class PropertyInfo {
```
[the remark this answers is not in the document]
Comment (jackwener) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/properties/DistributionSpecHash.java]: delete it

**Review by 924060929 — APPROVED** [review]
LGTM

**Pull request merged**